### PR TITLE
Add address in poi for autocomplete

### PR DESCRIPTION
**Pull request by kadhikari**
Add attribut address in poi. It can be of the type Addr or Street.
Ticket concerned: https://jira.canaltp.fr/browse/NAVP-721

### CODE REVIEW

**Comment by TeXitoi** [comment]
rustfmt fails

File: tests/bragi_poi_test.rs [hunk 1]
```diff
@@ -50,24 +50,24 @@ pub fn bragi_poi_test(es_wrapper: ::ElasticSearchWrapper) {
     // - osm_fixture.osm.pbf (including ways and pois)
     // - bano-three_cities
     // ******************************************
-    let osm2mimir = concat!(env!("OUT_DIR"), "/../../../osm2mimir");
+    let bano2mimir = concat!(env!("OUT_DIR"), "/../../../bano2mimir");
```
Comment (antoine-de) [inline tests/bragi_poi_test.rs]: the comment above is not synchronized

File: tests/bragi_poi_test.rs [hunk 2]
```diff
@@ -225,3 +227,31 @@ fn poi_misspelt_one_word_admin_test(bragi: &BragiHandler) {
...
+    assert!(count_types(&types, Poi::doc_type()) >= 1);
+    let poi = geocodings.first().unwrap();
+    assert_eq!(get_value(poi, "type"), Poi::doc_type());
+    assert_eq!(get_value(poi, "id"), "poi:osm:way:112361498");
+    assert_eq!(get_value(poi, "label"), "Hôtel de Ville (Melun)");
+    assert_eq!(get_poi_type_ids(poi), &["poi_type:amenity:townhall"]);
+    
+    assert_eq!(poi.get("address").and_then(|a| a.pointer("/Addr/house_number")), Some(&json!("2")));
+    assert_eq!(poi.get("address").and_then(|a| a.pointer("/Addr/id")), Some(&json!("addr:2.65801;48.53685")));
+    assert_eq!(poi.get("address").and_then(|a| a.pointer("/Addr/label")), Some(&json!("2 Rue de la Reine Blanche (Melun)")));
```
Comment (antoine-de) [inline tests/bragi_poi_test.rs]: I don't understand how the address has the city name in it, since there was no OSM at reading time :thinking:

File: libs/mimir/src/objects.rs [hunk 3]
```diff
@@ -52,6 +52,13 @@ pub enum Place {
     Stop(Stop),
 }
 
+/// Object stored in elastic search
+#[derive(Serialize, Deserialize, Debug, Clone)]
```
[the remark this answers is not in the document]
Comment (antoine-de) [reply libs/mimir/src/objects.rs]: hum I don't really see the implication of this

File: libs/mimir/src/rubber.rs [hunk 4]
```diff
@@ -90,6 +93,150 @@ pub fn get_date_index_name(base_index_name: &str) -> String {
...
+    pt_datasets: &[&str],
+    types: &[&str],
+    client: &mut rs_es::Client,
+) -> Result<Vec<String>, EsError> {
+    make_indexes_impl(all_data, pt_datasets, types, |index| {
+        is_existing_index(client, index)
+    })
+}
+
+pub fn make_indexes_impl<F: FnMut(&str) -> Result<bool, EsError>>(
```
Comment (antoine-de) [inline libs/mimir/src/rubber.rs]: I don't understand this code has been moved from somewhere no ?
I don't see where from in the PR

File: libs/mimir/src/rubber.rs [hunk 5]
```diff
@@ -90,6 +93,150 @@ pub fn get_date_index_name(base_index_name: &str) -> String {
...
+    }
+    match client.open_index(&index) {
+        //This error indicates that the search index is absent in ElasticSearch.
+        Err(EsError::EsError(_)) => Ok(false),
+        Err(e) => Err(e),
+        Ok(_) => Ok(true),
+    }
+}
+
+pub fn make_indexes(
```
Comment (antoine-de) [inline libs/mimir/src/rubber.rs]: it's not really an index creation there, isn't it `get_indexes` ?

File: libs/mimir/src/objects.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (TeXitoi) [reply libs/mimir/src/objects.rs]: Small, only on the elasticsearch database. May be relevant if we want to use that in the search later.

File: tests/bragi_poi_test.rs (hunk 2 above)
Comment (TeXitoi) [reply tests/bragi_poi_test.rs]: BANO has the city name, that's the list of administravite_regions that may be inexistent.

File: libs/mimir/src/rubber.rs (hunk 4 above)
Comment (TeXitoi) [reply libs/mimir/src/rubber.rs]: Looks like there is code duplication.

File: libs/mimir/src/rubber.rs [hunk 6]
```diff
@@ -90,6 +93,150 @@ pub fn get_date_index_name(base_index_name: &str) -> String {
...
+        "public_transport:stop_area" => "stop",
+        "city" => "admin",
+        "house" => "addr",
+        _ => a_type,
+    };
+
+    format!("munin_{}", doc_type)
+}
+
+pub fn collect(result: SearchResult<serde_json::Value>) -> Result<Vec<Place>, EsError> {
```
[the remark this answers is not in the document]
Comment (TeXitoi) [reply libs/mimir/src/rubber.rs]: To returns the same thing as an autocompletion.

File: libs/mimir/src/rubber.rs (hunk 4 above)
Comment (kadhikari) [reply libs/mimir/src/rubber.rs]: done

File: src/osm_reader/poi.rs [hunk 7]
```diff
@@ -288,6 +289,24 @@ pub fn compute_poi_weight(pois_vec: &mut [Poi], city_level: u32) {
...
+                    addr[0].address()
+                } else {
+                    None
+                }
+            }
+            _ => {
+                warn!("The poi {:?} {:?} doesn't have address", poi.id, poi.name);
+                None
+            }
+        };
```
Comment (TeXitoi) [inline src/osm_reader/poi.rs]:
```rust
poi.address = rubber
    .get_address(&poi.coord)
    .ok()
    .and_then(|addrs| addrs.into_iter().next())
    .map(|addr| addr.address());
if poi.address.is_none() {
    warn!("The poi {:?} {:?} doesn't have address", poi.id, poi.name);
}
```

File: tests/bragi_poi_test.rs (hunk 1 above)
Comment (TeXitoi) [reply tests/bragi_poi_test.rs]: still relevant: unsynchronized comment.

File: tests/bragi_poi_test.rs [hunk 8]
```diff
@@ -225,3 +226,39 @@ fn poi_misspelt_one_word_admin_test(bragi: &BragiHandler) {
...
+    // search poi: Poi as a relation in osm data
+    let geocodings = bragi.get("/autocomplete?q=Parking (Le Coudray-Montceaux)");
+    let types = get_types(&geocodings);
+    assert_eq!(count_types(&types, Poi::doc_type()), 1);
+    let first_poi = geocodings
+        .iter()
+        .find(|e| get_value(e, "type") == Poi::doc_type())
+        .unwrap();
+    assert_eq!(get_value(first_poi, "citycode"), "91179");
+    assert_eq!(get_poi_type_ids(first_poi), &["poi_type:amenity:parking"]);
```
Comment (TeXitoi) [inline tests/bragi_poi_test.rs]: That's just a copy and paste from another test? seems irrelevant here.

File: tests/bragi_poi_test.rs [hunk 9]
```diff
@@ -225,3 +226,39 @@ fn poi_misspelt_one_word_admin_test(bragi: &BragiHandler) {
...
+
+    // search poi: Poi as a way in osm data
```
Comment (TeXitoi) [inline tests/bragi_poi_test.rs]: We don't care of this detail here.

File: tests/bragi_poi_test.rs [hunk 10]
```diff
@@ -225,3 +226,39 @@ fn poi_misspelt_one_word_admin_test(bragi: &BragiHandler) {
...
+    let geocodings = bragi.get("/autocomplete?q=77000 Hôtel de Ville (Melun)");
+    let types = get_types(&geocodings);
+    assert!(count_types(&types, Poi::doc_type()) >= 1);
+    let poi = geocodings.first().unwrap();
+    assert_eq!(get_value(poi, "type"), Poi::doc_type());
+    assert_eq!(get_value(poi, "id"), "poi:osm:way:112361498");
+    assert_eq!(get_value(poi, "label"), "Hôtel de Ville (Melun)");
+    assert_eq!(get_poi_type_ids(poi), &["poi_type:amenity:townhall"]);
```
Comment (TeXitoi) [inline tests/bragi_poi_test.rs]: The two last asserts are not relevant here.

File: libs/mimir/src/rubber.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (antoine-de) [reply libs/mimir/src/rubber.rs]: I do not agree, it's nice that the API stays the same, but this is a helper, and it's not classy to do a `addrs.into_iter().next()` (or worse a check then `[0]` :stuck_out_tongue_winking_eye: ) to use the field.

I think this should return an `Option` and it should be wrapped in a `Vec` in the `GeocodingResponse`

[the remark this answers is not in the document]
Comment (TeXitoi) [reply libs/mimir/src/rubber.rs]: `fn collect` is also used in `fn features`. It get a raw elasticsearch response and gives you the corresponding `Vec<Place>`.

File: libs/mimir/src/rubber.rs [hunk 11]
```diff
@@ -258,6 +400,29 @@ impl Rubber {
             .collect())
     }
 
+    pub fn get_address(&mut self, coord: &Coord) -> Result<Vec<Place>, EsError> {
```
Comment (TeXitoi) [inline libs/mimir/src/rubber.rs]: @antoine-de this one can return an Option instead of a Vec, but not collect.

Comment (TeXitoi) [reply libs/mimir/src/rubber.rs]: It was like that before and changing this type will obscurate reverse. I don't know which one is better, but that's true that an Option reflect more the intent of this method.

File: libs/mimir/src/rubber.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (antoine-de) [reply libs/mimir/src/rubber.rs]: why does `features` returns a `Vec` ? (and collect is really not very explicit)

File: libs/mimir/src/rubber.rs (hunk 11 above)
Comment (antoine-de) [reply libs/mimir/src/rubber.rs]: ah yes maybe that's enough

File: libs/mimir/src/rubber.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (TeXitoi) [reply libs/mimir/src/rubber.rs]: features is used for autocomplete, thus a vector

[the remark this answers is not in the document]
Comment (TeXitoi) [reply libs/mimir/src/rubber.rs]: for the naming, that's Abder's fault! https://github.com/CanalTP/mimirsbrunn/commit/4dc9883d28429146454a980f08324d017663e16e

[the remark this answers is not in the document]
Comment (TeXitoi) [reply libs/mimir/src/rubber.rs]: my bad, feature is also a one response api. But that's not code from this PR.

@kadhikari feel free to refactor or not as this is not your responsability.

[the remark this answers is not in the document]
Comment (antoine-de) [reply libs/mimir/src/rubber.rs]: for the naming, that's not the same as at the naming time the function was private :stuck_out_tongue_winking_eye:

[the remark this answers is not in the document]
Comment (kadhikari) [reply libs/mimir/src/rubber.rs]: @TeXitoi for the refactor, i will wait monday since i won't  be present tomorrow

**Comment by kadhikari** [comment]
@TeXitoi could you merge as it is and i will midify in the next PR.
